service/history/api/workflow_id_dedup.go

```diff
@@ -100,6 +104,33 @@ func ResolveDuplicateWorkflowID(
 	return nil, nil
 }
 
+func resolveJustStartedDuplicatedWorfklows(
+	shardContext shard.Context,
+	currentWorkflowStartTime time.Time,
+	newRunID string,
+	workflowID string,
+) (UpdateWorkflowActionFunc, error) {
+	// we are going to use "grace period"
+	// to prevent multiple calls to start worklfows with the same ID
+	// if new workflow is starting earlier then that period - we will not terminate old worklfow,
+	// but rather didn't start the new one
+
+	gracePeriod := shardContext.GetConfig().WorkflowDeduplicationGracePeriod()
+	now := shardContext.GetTimeSource().Now()
+	if gracePeriod == 0 || now.After(currentWorkflowStartTime.Add(gracePeriod)) {
+		return terminateWorkflowAction(newRunID)
+	}
+
+	// there is a grace period, and old worklfow start time is within that period
+	// do not start new worklfow
+	msg := fmt.Sprintf("Too many restarts for worklfow %s with run ID %s", workflowID, newRunID)
```

Comment:
Probably should not include new runID because the new run is not created anyways. Maybe useful to include the time since existing run started, and the minimal interval we allow for WorkflowID to reuse.

Reply:
done. Added "time since workflow start" in milliseconds